### ISSUE TITLE
Parallel=methods not working when tests have different priorities set

### ISSUE THREAD

**Issue by infoneershalin** [cbeust/testng#1480]
### Expected behavior
Irrespective of priority, all testcases should run in parallel if executors are available.

### Actual behavior
It waits for higher priority testcase to complete and then starts testcase on another thread.

### Test case sample

This is sample class

```java
public class SimpleTest {
	@Test(priority = 1)
	public void test1() throws InterruptedException {
		System.out.println("Test 1:::Start::" + new Date());
		System.out.println("Test 1:::Thread::" + Thread.currentThread().getName());
		Thread.sleep(3000);
		System.out.println("Test 1:::End::" + new Date());

	}

	@Test(priority = 2)
	public void test2() throws InterruptedException {
		System.out.println("Test 2:::Start::" + new Date());
		System.out.println("Test 2:::Thread::" + Thread.currentThread().getName());
		Thread.sleep(3000);
		System.out.println("Test 2:::End::" + new Date());
	}
}
```

This is my config file
```xml
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Testng Issue" verbose="0" >
	<test name="1" enabled="true" parallel="methods">
		<classes>
			<class name="com.methodparalleltest.SimpleTest"></class>
		</classes>
	</test>
</suite>
```

Sample Output:

```
Test 1:::Start::Mon Jul 17 20:25:22 IST 2017
Test 1:::Thread::TestNG-test=1-1
Test 1:::End::Mon Jul 17 20:25:25 IST 2017
Test 2:::Start::Mon Jul 17 20:25:25 IST 2017
Test 2:::Thread::TestNG-test=1-2
Test 2:::End::Mon Jul 17 20:25:28 IST 2017
```

**Comment by juherr** [comment]
`priority` creates a dependency between methods and only the group of methods without depency can be run in parallel.

So, the actual behavior works as expected and I'm not sure we want to break the depency between tests when parallel is used.

**Comment by infoneershalin** [comment]
But for dependency creation there are dependsOnmethods and dependsOnGroups attributes.
So why we are taking priority also as dependency creation?

Do we have anyway right now through which I can ignore priority and execute tests in parallel?

Also, from which version this behavior has changed?


On 18-Jul-2017 2:42 AM, Julien Herr <notifications@github.com> wrote:

priority creates a dependency between methods and only the group of methods without depency can be run in parallel.

So, the actual behavior works as expected and I'm not sure we want to break the depency between tests when parallel is used.

**Comment by juherr** [comment]
> So why we are taking priority also as dependency creation?

Why do you not consider `priority` as a dependency description? It is just like `dependsOnX` but without the need to name the dependency.

> Do we have anyway right now through which I can ignore priority and execute tests in parallel?

You can put an `IAnnotationListener` which will unset the priority value.

> Also, from which version this behavior has changed?

Since 6.10 https://github.com/cbeust/testng/blob/master/CHANGES.txt#L77

**Comment by cjayswal** [comment]
I think Priority and dependency are two different things. Most of the time priority is used to order test or provide higher preference when there is competition. It doesn't mean dependency.

When we have limited resources/threads then priority should be considered to allocate resources/threads based on priority but If resources are available to serve we should not block request from low priority.

Consider the example that 5 threads are available
Usecase-1 
5 methods with priority 1,2,3,4,5
even if 5 threads are available only one threads will be utilized
Usecase-2
5 methods with priority 1,1,2,2,,3,3,4,4,4,5,5,5,5,5
it should run all 5 test in parallel because 5 threads are available (1,1,2,2,3)
it will use 2 threads for priority 1 other threads will remain ideal and other test will wait to complete priority 1 test! even if not dependent.

**Comment by juherr** [comment]
Right, `priority` and `dependency` are used to order tests.
But the current implementation doesn't manage strict/soft order.

Sadly, according to StackOverflow, many users are using `priority` as a dependency feature and will complain if we have regressions. 
So, I don't know how TestNG is supposed to manage parallelization with order features and if the order is strict or not:

| Order feature \ Parallel mode | methods | tests | classes | instances |
|-----------------|---------|-------|---------|-----------|
| groupByInstance |     ?    |   ?    |   ?      |   ?        |
| preserveOrder   |     ?    |    ?   |     ?    |      ?     |
| priority        |     ?    |   ?    |    ?     |     ?      |
| dependsOnGroups |     ?    |    ?   |     ?    |      ?     |
| dependsOnMethods |    ?     |    ?   |    ?     |    ?       |

Ping @cbeust WDYT?

**Comment by cbeust** [comment]
Correct, priorities and dependencies are different: if `a` has a lower priority than `b` and both methods are scheduled to be run, `a` will be run first. The important point here is that `b` is not required to be included in the run.

On the other hand, if `a` depends on `b`, then `a` will not run unless `b` has run first and has succeeded.

Another way of looking at it is that a priority is a soft dependency.

**Comment by juherr** [comment]
@cbeust Yes, I agree with your definition. But what about order execution?

If `a` has a lower priority than `b`, `a` will be run first and `b` will then be run: 

  1. Parallel or sequential?
    a. `b` may be run in the same time (parallel), or
    b. `b` must wait the end of `a`?
  2. Same or different thread?
    a. `b` may use another thread from the pool, or
    b. `b` must use the same thread than `a`?

Same question for each order features.

I feel the answer will depend on the parallel mode.

Could we try to answer the 2 questions for all use cases of the table from my previous comment according to the following sample?

```java
public class A {

  @Test
  public void testA() {}
}

public class B {

  @Test
  public void testB() {}
}
```
thread pool size: **>= 2**

**Comment by SaiPriyanka** [comment]
Any workaround for this issue

**Comment by krmahadevan** [comment]
@SaiPriyanka - We currently dont know if this is supposed to be treated as an issue or not. So there's no question of a workaround. But if you would like to nullify the effect of priorities and have all tests run in parallel then read @juherr comments : https://github.com/cbeust/testng/issues/1480#issuecomment-316045151

**Comment by krmahadevan** [comment]
@juherr 
While @cbeust gets back to your question, here's my opinion/thought on this

| Order feature \ Parallel mode | methods | tests | classes | instances | none |
|-------------------------------|---------|-------|---------|-----------|------|
| groupByInstance               | DT/EP   | DT/EP | DT/EP   | ST/NP     | RCB   |
| preserveOrder                 | ST/NP   | ST/NP | ST/NP   | ST/NP     | RCB   |
| priority                      | DT/EP   | DT/EP | DT/EP   | DT/EP     | RCB   |
| dependsOnGroups               | RCB     | RCB   | RCB     | RCB       | RCB  |
| dependsOnMethods              | RCB     | RCB   | RCB     | RCB       | RCB  |

* DT/EP - Different Threads/Enable Parallel execution
* ST/NP - Single Thread/No Parallel execution
* RCB - Retain Current Behavior

**Comment by juherr** [comment]
@krmahadevan Could you specify what is the current behavior?

**Comment by krmahadevan** [comment]
Sure. I will do that.

**Comment by MicahLC** [comment]
Has there been any movement on this? Per @cjayswal's comment at https://github.com/cbeust/testng/issues/1480#issuecomment-321059611, this gives very inefficient utilization of test threads. I'd like to have priorities (I have slow tests that I would like to run first) AND still maximize thread utilization.

**Comment by MicahLC** [comment]
Also, this appears to have been the behavior since 6.9.11, per https://github.com/cbeust/testng/blob/master/CHANGES.txt#L276. I've been able to work around this for now by using 6.9.10.

**Comment by juherr** [comment]
@MicahLC  Did you try with a recent version? Because it changed a lot with 6.10 and later.

**Comment by MicahLC** [comment]
@juherr I just tried it with 6.14.3 and still have issues.

**Comment by MicahLC** [comment]
I just created a PR to add a very simple test case that I think exemplifies this behavior. It uses one suite with two tests, each of which has one class, but I've seen the behavior with just one test with one class. Lower-priority methods that don't have any dependencies don't even get **started** executing until the higher-priority methods in front of them **finish executing**, which is extremely inefficient.
